File: Data Structures & Algorithms/reverse-a-linked-list/submission-1.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if not head or head.next == None :
            return head
        new_list_head = ListNode(-10001)
        curr = head
        while curr:
            new_node = ListNode(curr.val)
            new_node.next = new_list_head
            new_list_head = new_node
            curr = curr.next
        curr = new_list_head
        while curr.next.val != -10001:
            curr = curr.next
        curr.next = None
        return new_list_head

            

File: Data Structures & Algorithms/reorder-linked-list/submission-1.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    void reorderList(ListNode* head) {
        ListNode* slow = head;
        ListNode* fast= head;

        while(fast && fast->next){
            slow = slow->next;
            fast = fast->next->next;
        }

        ListNode* head2 = slow->next;
        slow->next = nullptr;

        ListNode* tail = head2;
        ListNode* prev = nullptr;
        while(tail){
            ListNode* temp = tail->next;
            tail->next = prev;
            prev = tail;
            tail = temp;
        }

        head2 = prev;
        ListNode* first = head;
        ListNode* second = head2;
        while(second){
            ListNode* temp1 = first->next;
            ListNode* temp2 = second->next;

            first->next = second;
            second->next = temp1;

            first = temp1;
            second = temp2;

        }

    }
};
